pygfx/renderers/wgpu/_mipmapsutil.py

```diff
@@ -54,155 +57,144 @@
-                base_mip_level=i,
-                mip_level_count=1,
-                dimension="2d",
-                base_array_layer=base_array_layer,
-            )
+        bind_group = device.create_bind_group(
+            layout=bind_group_layout,
+            entries=[
+                {"binding": 0, "resource": src_view},
+                {"binding": 1, "resource": dst_view},
+            ],
+        )
 
-            pass_encoder = command_encoder.begin_compute_pass()
+        pass_encoder.set_pipeline(pipeline)
+        pass_encoder.set_bind_group(0, bind_group, [], 0, 99)
+        pass_encoder.dispatch_workgroups(dst_size[0], dst_size[1])
+        pass_encoder.end()
 
-            bind_group = self.device.create_bind_group(
-                layout=bind_group_layout,
-                entries=[
-                    {"binding": 0, "resource": src_view},
-                    {"binding": 1, "resource": dst_view},
-                ],
-            )
+        src_view = dst_view
 
-            pass_encoder.set_pipeline(pipeline)
-            pass_encoder.set_bind_group(0, bind_group, [], 0, 99)
-            pass_encoder.dispatch_workgroups(dst_size[0], dst_size[1])
-            pass_encoder.end()
+    device.queue.submit([command_encoder.finish()])
 
-            src_view = dst_view
 
-        self.device.queue.submit([command_encoder.finish()])
+def get_mipmap_pipeline(device, texture):
+    format = texture._wgpu_object.format
+    pipeline = MIPMAP_CACHE.get(format)
 
+    if pipeline is None:
+        if format.endswith("srgb"):
+            raise RuntimeError("Cannot create mipmaps for srgb textures.")
 
-_the_mipmap_util = None
+        type, textype = "f32", "float"
+        if format.endswith(("norm", "float")):
+            type, textype = "f32", "float"
+        elif format.endswith("sint"):
+            type, textype = "i32", "sint"
+        elif format.endswith("uint"):
+            type, textype = "u32", "uint"
+
+        shader = mipmap_source.replace("FORMAT", format).replace("TYPE", type)
+        module = device.create_shader_module(code=shader)
+
+        pipeline = device.create_compute_pipeline(
+            layout=create_pipeline_layout(device, format, textype),
+            compute={
+                "module": module,
+                "entry_point": "c_main",
+            },
+        )
 
+        MIPMAP_CACHE.set(format, pipeline)
 
-def generate_texture_mipmaps(device, target):
-    # If this looks like a cube or stack, generate mipmaps for each individual layer
-    if isinstance(target, Texture) and target.dim == 2 and target.size[2] > 1:
-        for i in range(target.size[2]):
-            generate_texture_mipmaps(
-                device, GfxTextureView(target, layer_range=(i, i + 1))
-            )
-        return
+    # Strore a ref of the pipeline, the cache uses weak refs.
+    # I.e. the pipeline for this format is in the cache for as long as
+    # any textures that use it are alive.
```

Comment:
I'm wondering, if the possible number of pipelines is low (e.g. there are only X possible pipelines to be generated, let's say X < 10), why would you ever evict them from cache?

Reply:
That's a fair point in this case. But in other cases (that I plan to implement later) this will matter. I reckoned that its better to be consistent from the start.